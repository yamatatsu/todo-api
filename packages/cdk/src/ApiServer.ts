import {
  App,
  Stack,
  StackProps,
  aws_ec2,
  aws_lambda,
  aws_lambda_nodejs,
  aws_apigateway,
  aws_secretsmanager,
  Duration,
  aws_cognito,
} from "aws-cdk-lib";

type Props = StackProps & {
  codeEntry: string;
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
  dbCredentialSecret: aws_secretsmanager.ISecret;
  proxyEndpoint: string;
  userPool: aws_cognito.IUserPool;
};

export class ApiServerStack extends Stack {
  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    // TODO: lambda insight入れる
    const handler = new aws_lambda_nodejs.NodejsFunction(this, "Lambda", {
      entry: props.codeEntry,
      runtime: aws_lambda.Runtime.NODEJS_14_X,
      // TODO: aurora serverless やめて rds proxyつけたら10秒に戻す
      timeout: Duration.seconds(20),
      vpc: props.vpc,
      securityGroups: [props.securityGroup],
      bundling: {
        /**
         * esbuildの成果物にprismaのengineが含める必要がある
         * Respect for https://dev.to/prisma/bundling-prisma-with-the-cdk-using-aws-lambda-nodejs-2lkd
         */
        nodeModules: ["@prisma/client", "prisma"],
        commandHooks: {
          beforeBundling(inputDir: string, outputDir: string): string[] {
            return [];
          },
          beforeInstall(inputDir: string, outputDir: string) {
            return [`cp -R ../prisma ${outputDir}/`];
          },
          afterBundling(inputDir: string, outputDir: string): string[] {
            return [
              `cd ${outputDir}`,
              `yarn prisma generate`,
              `rm -rf node_modules/@prisma/engines`,
            ];
          },
        },
        /**
         * forceDockerBundling: これは default false であるが、下記理由により明示的に示す。
         * これが true のとき、もしくは esbuild が見つからないとき(下記URL参照)に、
         * NodejsFunction では docker を用いてビルドが行われる。
         * https://github.com/aws/aws-cdk/blob/4c8e938e01b87636390a4f04de63bcd4dfe44cf8/packages/@aws-cdk/aws-lambda-nodejs/lib/esbuild-installation.ts#L8-L31
         * dockerでビルドする場合、特に Mac M1 での実行結果と GitHub actions 上での実行結果とで、
         * 成果物のfingerprintがズレて snapshot test が落ちてしまう。
         * そのため、 forceDockerBundling を明示し、 package.json でも esbuild を明示的にインストールする。
         */
        forceDockerBundling: false,
      },
      environment: {
        NODE_ENV: "production",
        SECRET_NAME: props.dbCredentialSecret.secretName,
        DATABASE_ENDPOINT: props.proxyEndpoint,
      },
    });

    props.dbCredentialSecret.grantRead(handler);

    // const authorizer = new aws_apigateway.CognitoUserPoolsAuthorizer(
    //   this,
    //   "Authorizer",
    //   {
    //     cognitoUserPools: [],
    //     identitySource: "method.request.header.x-ta-token",
    //   }
    // );

    const restApi = new aws_apigateway.LambdaRestApi(this, "LambdaRestApi", {
      handler,
      deployOptions: {
        metricsEnabled: true,
        loggingLevel: aws_apigateway.MethodLoggingLevel.INFO,
        dataTraceEnabled: true,
        tracingEnabled: true,
      },
      // defaultMethodOptions: {
      //   authorizer,
      // },
    });
  }
}
