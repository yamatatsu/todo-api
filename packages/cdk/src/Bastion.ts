import { App, Stack, StackProps, aws_ec2, aws_iam } from "aws-cdk-lib";

type BastionProps = {
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
} & StackProps;

export class BastionStack extends Stack {
  readonly bastion: aws_ec2.Instance;

  constructor(scope: App, id: string, props: BastionProps) {
    super(scope, id, props);

    const instanceType = new aws_ec2.InstanceType("t2.micro");
    const machineImage = new aws_ec2.AmazonLinuxImage({
      generation: aws_ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
    });

    // ec2 instance for bastion server.
    this.bastion = new aws_ec2.Instance(this, "Bastion", {
      instanceType: instanceType,
      machineImage: machineImage,
      vpc: props.vpc,
      vpcSubnets: { subnetGroupName: "application" },
    });

    this.bastion.addSecurityGroup(props.securityGroup);

    // for ssm session manager
    this.bastion.addToRolePolicy(
      new aws_iam.PolicyStatement({
        effect: aws_iam.Effect.ALLOW,
        actions: [
          "ssm:UpdateInstanceInformation",
          "ssmmessages:CreateControlChannel",
          "ssmmessages:CreateDataChannel",
          "ssmmessages:OpenControlChannel",
          "ssmmessages:OpenDataChannel",
        ],
        resources: ["*"],
      })
    );

    this.bastion.addToRolePolicy(
      new aws_iam.PolicyStatement({
        effect: aws_iam.Effect.ALLOW,
        actions: ["s3:GetEncryptionConfiguration"],
        resources: ["*"],
      })
    );
  }
}
