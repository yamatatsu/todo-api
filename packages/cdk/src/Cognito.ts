import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  aws_cognito,
} from "aws-cdk-lib";

export class CognitoStack extends Stack {
  public readonly userPool: aws_cognito.IUserPool;

  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props);

    const userPool = new aws_cognito.UserPool(this, "UserPool", {
      selfSignUpEnabled: true,
      signInAliases: { username: true, email: true },
      customAttributes: {
        validated_email: new aws_cognito.StringAttribute({
          mutable: true,
        }),
      },
      passwordPolicy: {
        // Because of https://pages.nist.gov/800-63-3/sp800-63b.html#-5112-memorized-secret-verifiers
        minLength: 8,
        // Because of https://pages.nist.gov/800-63-3/sp800-63b.html#a3-complexity
        requireLowercase: false,
        requireUppercase: false,
        requireDigits: false,
        requireSymbols: false,
      },
      removalPolicy: RemovalPolicy.DESTROY,
    });

    new aws_cognito.UserPoolClient(this, "UserPoolClient", {
      userPool,
      authFlows: { userSrp: true },
    });

    this.userPool = userPool;
  }
}
