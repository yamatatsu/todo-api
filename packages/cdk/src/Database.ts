import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  aws_ec2,
  aws_rds,
  aws_secretsmanager,
} from "aws-cdk-lib";

type Props = StackProps & {
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
};

export class DatabaseStack extends Stack {
  public readonly dbCredentialSecret: aws_secretsmanager.ISecret;

  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    const dbCluster = new aws_rds.ServerlessCluster(this, "Aurora", {
      engine: aws_rds.DatabaseClusterEngine.AURORA,
      credentials: aws_rds.Credentials.fromGeneratedSecret("LambdaRdsProxy"),
      defaultDatabaseName: "tadb",
      vpc: props.vpc,
      vpcSubnets: { subnetType: aws_ec2.SubnetType.ISOLATED },
      securityGroups: [props.securityGroup],
      removalPolicy: RemovalPolicy.DESTROY,
    });

    // production codeではないので、カジュアルに`!`使う
    this.dbCredentialSecret = dbCluster.secret!;
  }
}
