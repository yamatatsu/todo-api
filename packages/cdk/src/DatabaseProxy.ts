import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  aws_ec2,
  aws_rds,
  aws_secretsmanager,
} from "aws-cdk-lib";

type Props = StackProps & {
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
  cluster: aws_rds.IDatabaseCluster;
  secret: aws_secretsmanager.ISecret;
};

export class DatabaseProxyStack extends Stack {
  public readonly proxy: aws_rds.IDatabaseProxy;

  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    const proxy = props.cluster.addProxy("RdsProxy", {
      vpc: props.vpc,
      secrets: [props.secret],
      securityGroups: [props.securityGroup],
    });
    // proxy.connections.allowFromAnyIpv4(aws_ec2.Port.tcp(3306));

    this.proxy = proxy;
  }
}
