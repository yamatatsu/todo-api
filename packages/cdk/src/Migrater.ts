import {
  App,
  Stack,
  StackProps,
  aws_ec2,
  aws_lambda,
  aws_secretsmanager,
  Duration,
} from "aws-cdk-lib";

type Props = StackProps & {
  dockerfilePath: string;
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
  dbCredentialSecret: aws_secretsmanager.ISecret;
  proxyEndpoint: string;
};

/**
 * Database に対して prisma migrate deploy をかけるための lambda 関数を定義する。
 * application subnet 内で実行されることで、databaseへアクセス可能となる。
 */
export class MigraterStack extends Stack {
  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    const handler = new aws_lambda.DockerImageFunction(this, "Lambda", {
      functionName: `${id}-Fn`,
      code: aws_lambda.DockerImageCode.fromImageAsset(props.dockerfilePath),
      // 途中で実行止まると怖いので5分くらいを設定する
      timeout: Duration.minutes(5),
      vpc: props.vpc,
      vpcSubnets: { subnetGroupName: "application" },
      securityGroups: [props.securityGroup],
      environment: {
        SECRET_NAME: props.dbCredentialSecret.secretName,
        DATABASE_ENDPOINT: props.proxyEndpoint,
      },
    });

    props.dbCredentialSecret.grantRead(handler);
  }
}
