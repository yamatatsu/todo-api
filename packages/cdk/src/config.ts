const ENV_NAME = ["development", "production"] as const;
type EnvName = typeof ENV_NAME[number];

export function assertEnvName(str: string): asserts str is EnvName {
  if (!ENV_NAME.includes(str as EnvName)) {
    throw new Error(`Invalid EnvName has set: ${str}`);
  }
}

export const AccountId: Record<EnvName, string> = {
  development: "407172421073",
  production: "",
};
