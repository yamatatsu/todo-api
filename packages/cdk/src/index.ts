import path from "path";
import { App, Environment } from "aws-cdk-lib";
import { CognitoStack } from "./Cognito";
import { VpcStack } from "./Vpc";
import { DatabaseStack } from "./Database";
import { ApiServerStack } from "./ApiServer";
import { BastionStack } from "./Bastion";
import { AccountId, assertEnvName } from "./config";

const envName = process.env.ENV_NAME || "development";
assertEnvName(envName);

const app = new App();
const env: Environment = {
  region: "ap-northeast-1",
  account: AccountId[envName],
};

const cognito = new CognitoStack(app, `${envName}-TodoApi-Cognito`, { env });
const { vpc, dbSG, dbAccessSG } = new VpcStack(app, `${envName}-TodoApi-Vpc`, {
  env,
});
const database = new DatabaseStack(app, `${envName}-TodoApi-Database`, {
  vpc,
  securityGroup: dbSG,
  env,
});
new ApiServerStack(app, `${envName}-TodoApi-ApiServer`, {
  codeEntry: path.resolve(__dirname, "../../server/src/index.ts"),
  vpc,
  securityGroup: dbAccessSG,
  dbCredentialSecret: database.dbCredentialSecret,
  userPoolArn: cognito.userPoolArn,
  env,
});
new BastionStack(app, `${envName}-TodoApi-Bastion`, {
  vpc,
  securityGroup: dbAccessSG,
  env,
});
