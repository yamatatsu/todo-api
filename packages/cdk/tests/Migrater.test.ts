import path from "path";
import { SynthUtils } from "@aws-cdk/assert";
import {
  App,
  aws_cognito,
  aws_ec2,
  aws_secretsmanager,
  Stack,
} from "aws-cdk-lib";
import { MigraterStack } from "../src/Migrater";

test("snapshot test", () => {
  const app = new App();
  const stack = new Stack(app, "dummyStack");

  const vpc = new aws_ec2.Vpc(stack, "Vpc", {
    subnetConfiguration: [
      { name: "application", subnetType: aws_ec2.SubnetType.ISOLATED },
      { name: "rds", subnetType: aws_ec2.SubnetType.ISOLATED },
    ],
  });
  const securityGroup = new aws_ec2.SecurityGroup(stack, "SecurityGroup", {
    vpc,
  });
  const secret = new aws_secretsmanager.Secret(stack, "Secret");

  const target = new MigraterStack(app, "Target", {
    dockerfilePath: path.resolve(__dirname, "./dummy"),
    vpc,
    securityGroup,
    dbCredentialSecret: secret,
    proxyEndpoint: "test-proxyEndpoint",
  });

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot();
});
