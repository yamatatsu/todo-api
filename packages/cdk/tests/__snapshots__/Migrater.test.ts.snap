// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Resources": Object {
    "LambdaD247545B": Object {
      "DependsOn": Array [
        "LambdaServiceRoleDefaultPolicyDAE46E21",
        "LambdaServiceRoleA8ED4D3B",
      ],
      "Properties": Object {
        "Code": Object {
          "ImageUri": Object {
            "Fn::Sub": "\${AWS::AccountId}.dkr.ecr.\${AWS::Region}.\${AWS::URLSuffix}/cdk-hnb659fds-container-assets-\${AWS::AccountId}-\${AWS::Region}:4b876f80f5f7d09e3c026c3904b4c5ef7b6eb18b1602e8e52147695595eb474c",
          },
        },
        "Environment": Object {
          "Variables": Object {
            "DATABASE_ENDPOINT": "test-proxyEndpoint",
            "SECRET_NAME": Object {
              "Fn::Join": Array [
                "-",
                Array [
                  Object {
                    "Fn::Select": Array [
                      0,
                      Object {
                        "Fn::Split": Array [
                          "-",
                          Object {
                            "Fn::Select": Array [
                              6,
                              Object {
                                "Fn::Split": Array [
                                  ":",
                                  Object {
                                    "Fn::ImportValue": "dummyStack:ExportsOutputRefSecretA720EF052D953DED",
                                  },
                                ],
                              },
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  Object {
                    "Fn::Select": Array [
                      1,
                      Object {
                        "Fn::Split": Array [
                          "-",
                          Object {
                            "Fn::Select": Array [
                              6,
                              Object {
                                "Fn::Split": Array [
                                  ":",
                                  Object {
                                    "Fn::ImportValue": "dummyStack:ExportsOutputRefSecretA720EF052D953DED",
                                  },
                                ],
                              },
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              ],
            },
          },
        },
        "FunctionName": "Target-Fn",
        "PackageType": "Image",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaServiceRoleA8ED4D3B",
            "Arn",
          ],
        },
        "Timeout": 300,
        "VpcConfig": Object {
          "SecurityGroupIds": Array [
            Object {
              "Fn::ImportValue": "dummyStack:ExportsOutputFnGetAttSecurityGroupDD263621GroupIdDF6F8B09",
            },
          ],
          "SubnetIds": Array [
            Object {
              "Fn::ImportValue": "dummyStack:ExportsOutputRefVpcapplicationSubnet1Subnet8B68205B1F66BB2C",
            },
            Object {
              "Fn::ImportValue": "dummyStack:ExportsOutputRefVpcapplicationSubnet2Subnet296709622ABA6FC7",
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaServiceRoleA8ED4D3B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaServiceRoleDefaultPolicyDAE46E21": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::ImportValue": "dummyStack:ExportsOutputRefSecretA720EF052D953DED",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaServiceRoleDefaultPolicyDAE46E21",
        "Roles": Array [
          Object {
            "Ref": "LambdaServiceRoleA8ED4D3B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
